File: 2023/go/day06/src/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

const training string = `Time:      7  15   30
Distance:  9  40  200`

func main() {
	data := training
	data = getFileContent("src/input.txt")
	times, distances := parseData(data)
	fmt.Printf("times: %v, distances: %v\n", times, distances)
	computePart1(times, distances)
}

func parseData(data string) ([]int, []int) {
	split := strings.Split(data, "\n")
	times := getInt(split[0])
	distances := getInt(split[1])
	return times, distances
}

func getInt(line string) []int {
	ints := make([]int, 0)
	line = strings.Replace(line, " ", "", -1)
	for _, chunk := range strings.Split(line, ":") {
		if integer, err := strconv.Atoi(chunk); err == nil {
			ints = append(ints, integer)
		}
	}
	return ints
}

func computePart1(times, distances []int) {
	product := 1
	for i := 0; i < len(times); i++ {
		count := 0
		for speed := 0; speed < times[i]; speed++ {
			if speed*(times[i]-speed) > distances[i] {
				count++
			}
		}
		product *= count
	}
	fmt.Printf("Product: %v\n", product)
}

func getFileContent(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		panic(err)
	}
	return string(data)
}

File: 2023/rust/day06/src/main.rs
const _TRAINING: &str = "Time:      7  15   30
Distance:  9  40  200";

fn main() {
    // let data = _TRAINING;
    let data = include_str!("input.txt");
    let (times, distances) = parse_data(data);
    println!("times: {:?}, distances: {:?}", times, distances);
    let part1 = compute_games(times, distances);
    println!("Part one: {}", part1);
}

fn parse_data(data: &str) -> (Vec<usize>, Vec<usize>) {
    let mut split = data.split("\n");
    let (times, distances) = (
        get_ints(split.next().unwrap()),
        get_ints(split.next().unwrap()),
    );
    (times, distances)
}

fn get_ints(data: &str) -> Vec<usize> {
    data.replace(" ", "")
        .split(":")
        .skip(1)
        .map(|x| {
            if let Ok(int) = x.parse::<usize>() {
                return Some(int);
            }
            None
        })
        .filter(|option| option.is_some())
        .map(|x| x.unwrap())
        .collect::<Vec<usize>>()
}

fn compute_games(times: Vec<usize>, distances: Vec<usize>) -> usize {
    times
        .iter()
        .zip(distances.iter())
        .map(|(time, distance)| {
            (0..*time)
                .map(|speed| speed * (*time - speed))
                .filter(|score| score > distance)
                .count()
        })
        .product::<usize>()
}
